src/pymor/models/interface.py

```diff
@@ -109,6 +109,86 @@ def _compute_output(self, solution, mu=None, **kwargs):
         else:
             return self.output_functional.apply(solution, mu=mu).to_numpy()
 
+    def _compute_solution_d_mu_single_direction(self, parameter, index, solution, mu=None, **kwargs):
+        """Compute the partial derivative of the solution w.r.t. a parameter index
+
+        Parameters
+        ----------
+        parameter
+            parameter for which to compute the sensitivity
+        index
+            parameter index for which to compute the sensitivity
+        solution
+            Internal model state for the given |Parameter value|.
+        mu
+            |Parameter value| for which to solve
+
+        Returns
+        -------
+        The sensitivity of the solution as a |VectorArray|.
+        """
+        raise NotImplementedError
+
+    def _compute_solution_d_mu(self, solution, mu=None, **kwargs):
+        """Compute all partial derivative of the solution w.r.t. a parameter index
+
+        Parameters
+        ----------
+        solution
+            Internal model state for the given |Parameter value|.
+        mu
+            |Parameter value| for which to solve
+
+        Returns
+        -------
+        A dict of all partial sensitivities of the solution.
+        """
+        sensitivities = {}
+        for (parameter, size) in self.parameters.items():
+            sens_for_param = self.solution_space.empty()
+            for l in range(size):
+                sens_for_param.append(self._compute_solution_d_mu_single_direction(
+                    parameter, l, solution, mu))
+            sensitivities[parameter] = sens_for_param
+        return sensitivities
+
+    def _compute_output_d_mu(self, solution, mu=None, return_array=False, **kwargs):
+        """compute the gradient w.r.t. the parameter of the output functional
+
+        Parameters
+        ----------
+        solution
+            Internal model state for the given |Parameter value|.
+        mu
+            |Parameter value| for which to compute the gradient
+
+        Returns
+        -------
+        The gradient as a numpy array
+        """
+        U_d_mus = self._compute_solution_d_mu(solution, mu)
+        gradient_for_outputs = []
+        for d in range(self.output_functional.range.dim):
+            gradient = [] if return_array else {}
+            for (parameter, size) in self.parameters.items():
+                list_for_param = []
+                for index in range(size):
+                    output_partial_dmu = self.output_functional.d_mu(parameter, index).apply(
+                        solution, mu=mu).to_numpy()[0,d]
+                    U_d_mu = U_d_mus[parameter][index]
+                    list_for_param.append(output_partial_dmu
+                                + self.output_functional.jacobian(solution, mu).apply(
+                        U_d_mu, mu).to_numpy()[0,d])
```

Comment:
If I understand this correctly, you now produce a flat list for each parameter. I think what we discussed would mean to have a 2d-array per parameter where the first index is the index of the parameter and the second index the output component.